### ISSUE TITLE
Ichigo Training (Issue: )

### ISSUE THREAD

**Issue by hahuyhoang411** [janhq/ichigo#122]

**Comment by hahuyhoang411** [comment]
Lora for pretraining ref: https://unsloth.ai/blog/contpretraining

**Comment by tuanlda78202** [comment]
Paired with @bachvudinh to understand the fully trained original Ichigo pipeline.

**Comment by dan-homebrew** [comment]
@hahuyhoang411 - Can we combine this into #116

**Comment by bachvudinh** [comment]
## Data Quality Issue Resolution and Pipeline Update

### Issue
Text-to-semantic model undertrained on English, causing noise in small fraction of output data.

### Analysis
- instruction-speech-v1 and instruction-speech-v2 subsets showed highest percentage of affected data
- Using old data pipeline to regenerted the data: Text → Audio (WhisperSpeech TTS) → Semantic tokens (Ichigo Tokenizer)

### Resolution
- Filtered outlier noise data
- Regenerated data for instruction-speech-v1 and instruction-speech-v2 subsets -- in the end, the number of backfilled data is around  4.96k samples

## Quick link
- FIltered data: https://huggingface.co/datasets/homebrewltd/Ichigo-instruction-tokenized-v0.2-clean.
- Backfilled : https://huggingface.co/datasets/homebrewltd/Ichigo-instruction-tokenized-v0.2-bug.

**Comment by bachvudinh** [comment]
## Goal
- Ichigo can understand Vietnamese and English.

## Methodology
- To be updated

## Hyperparams
Parameter | Value
--- | ---
Rank | Vary from 2-64
Alpha | 
Epoch | 1
Global batch size | 32
Learning Rate | 8e-6
Learning Scheduler | Cosine
Optimizer | AdamW Fused
Warmup Steps | 7
Weight Decay | 0.01
Gradient Checkpointing | Full
Max length | 4096
Precision | bf16

## Results
- Text-only benchmark:
   + MMLU:
   + MMLU pro:  
   + VMLU: 
     
- Audio-Bench:
  + Speech Instruction (GPT-4 judge):
    Alpaca : 
    Open-hermes: 
    
  
  + ASR (WER score):
    

## Learnings
- Model return repetition response when processing the low quality audio

## Quicklinks
- Data: 
  + Speech instruction data v0.5: https://huggingface.co/datasets/homebrewltd/Ichigo-instruction-tokenized-v0.2-clean.

**Comment by bachvudinh** [comment]
## Goal
- Instruction tuning phase of Ichigo training.

## Methodology
- Instruction fine-tuning Llama 3.1 8B model with an expanded vocabulary of 2,560 sound semantic tokens.
- Data methodology: same as Ichigo v0.4 but removing noise-rejection data.

## Hyperparams
Parameter | Value
--- | ---
Epoch | 1
Global batch size | 256
Learning Rate | 3e-4
gradient_clipping | 1.0
Learning Scheduler | Cosine
Optimizer | AdamW Fused
Warmup Steps | 100
Weight Decay | 0.01
Gradient Checkpointing | Full
Max length | 4096
Precision | bf16
compile | True


## Results
- Time estimate: Took around ~32 hours on 8xH100 (runpod).
- Text-only benchmark:
   + MMLU:
   + MMLU pro:  
   + VMLU: 
     
- Audio-Bench:
  + Speech Instruction (GPT-4 judge):
    Alpaca : 
    Open-hermes: 
    
  
  + ASR (WER score):
    

## Learnings
- Model return repetition response when processing the low quality audio

## Quicklinks
- Data: 
  + Speech instruction data v0.5: https://huggingface.co/datasets/homebrewltd/Ichigo-instruction-tokenized-v0.2-clean. 
- Checkpoint:
  + Pretrain checkpoint: https://huggingface.co/jan-hq/ichigo-v0.5-llama3.2-8B-base.

**Comment by bachvudinh** [comment]
## Data Quality Issue Resolution and Pipeline Update

### Issue
Text-to-semantic model undertrained on English, causing noise in small fraction of output data.

### Analysis
- instruction-speech-v1 and instruction-speech-v2 subsets showed highest percentage of affected data
- Using old data pipeline to regenerted the data: Text → Audio (WhisperSpeech TTS) → Semantic tokens (Ichigo Tokenizer)

### Resolution
- Filtered outlier noise data
- Regenerated data for instruction-speech-v1 and instruction-speech-v2 subsets -- in the end, the number of backfilled data is around  4.96k samples

## Quick link
- FIltered data: https://huggingface.co/datasets/homebrewltd/Ichigo-instruction-tokenized-v0.2-clean.
- Backfilled : https://huggingface.co/datasets/homebrewltd/Ichigo-instruction-tokenized-v0.2-bug.

**Comment by hahuyhoang411** [comment]
Can you add an example of affected data @bachvudinh

**Comment by bachvudinh** [comment]
# Instruction tuning phase of Ichigo v0.5(Second Attempt)

## Methodology
- Instruction fine-tuning the pretrained checkpoint.
- Data methodology: Fixing the quality of viettel x nvidia data. see https://github.com/janhq/ichigo-internal/issues/8 for details.

## Hyperparams
Parameter | Value
--- | ---
Epoch | 1
Global batch size | 256
Learning Rate | 3e-4
gradient_clipping | 1.0
Learning Scheduler | Cosine
Optimizer | AdamW Fused
Warmup Steps | 100
Weight Decay | 0.01
Gradient Checkpointing | Full
Max length | 4096
Precision | bf16
compile | True

## Results
- Time estimate: Took around ~5 hours on 8xH100 (runpod).
- Loss Curve: 
  
   ![image](https://github.com/user-attachments/assets/7d8a141b-ce12-4434-a957-1f1505a38e2a)

- Instruction-Following Evaluation  
  | Benchmark            | Category                   | Metric                     | Score     |
  | :------------------- | :------------------------- | :------------------------- | :-------- |
  | **Text-only**        | MMLU                       |                           | **61.74** |
  |                      | MMLU pro                   |                           |           |
  |                      | VMLU                       |                           |           |
  | **Audio-Bench**      | Alpaca(Instruction Speech) | GPT-4 judge                    |           |
  |                      |    Open-hermes(Instruction Speech)                | GPT-4 judge               |           |
  |                      | ASR                        | WER score                  |    Not benchmarking yet due to early stop.      |
    
  
    

## Learnings
- The loss declined significantly and stabilized between 0.78-0.8.

## Quicklinks
- Data: 
  + Speech instruction data v0.5: https://huggingface.co/datasets/homebrewltd/Ichigo-instruction-tokenized-v0.2-clean. 
- Checkpoint:

**Comment by bachvudinh** [comment]
Some very good result from Ichigo v0.5 checkpoint step 4000. The response of model is now clearer and more helpful: 

![image](https://github.com/user-attachments/assets/7c42f122-c81e-42a5-8dec-76d3274888fd)

![image](https://github.com/user-attachments/assets/aa853e4a-64d6-4973-a376-027dff1fb1d4)

![image](https://github.com/user-attachments/assets/18580d69-0710-49b7-ba54-34d25ce1e764)

**Comment by hahuyhoang411** [comment]
It's also good on my end. But seems like it's too noise sensitive
<img width="1026" alt="Screenshot 2025-01-14 at 19 26 59" src="https://github.com/user-attachments/assets/dcd88b25-dc53-4f46-84ab-5996a45b6e76" />
<img width="1023" alt="Screenshot 2025-01-14 at 19 26 28" src="https://github.com/user-attachments/assets/c62637d0-5dcd-4605-a39f-040d64eebb8d" />

**Comment by bachvudinh** [comment]
# Training Issues and Budget Request Report
We are currently training a large language model using 8x H100 GPUs via RunPod. The initial budget estimation was $800 for a 32-hour training run.

## Technical Issues Encountered
- Initial 2000 training steps (~$150) were lost due to quality of viettel x nvidia data: https://github.com/janhq/ichigo-internal/issues/8.
- During the training process, we encountered an unexpected loss explosion consistently occurring around step 6000. Initial hypothesis focused on learning rate optimization, leading to multiple experimental runs to tune this parameter. --> Root cause analysis eventually identified a corrupted batch in the training data.

## Financial Impact
- Original Budget: $800 (32 hours).
- Viettel x Nvidia data bug: ~$150
- Debugging Costs: ~$200 (learning rate optimization attempts).
- Current Status: Training duration has exceeded initial estimates due to necessary debugging and data cleanup.

**Comment by tikikun** [comment]
## Issues report


We are facing some issues on ichigo v0.5, in short, it is really bad:

### Issues:
- The model can no longer follow system prompt (it seems continual pretraining effect has been broken due to configuration change somehow)
- Model quality in speech is not good enough ( @bachvu02 please link the speech benchmark degraded quality not much but still)


### Root causes:
- We have 5% of the data having error (missing <|end_token|>)
- Configuration change (possibly) learning rate is being changed due to multiple issues during the training, hence it seems damaging the "recovering" process (which is not continual but more overfit on the dataset at hands)
- Data quality just suck (we realized this during the training but continued anyways)

### Mitigation plans:
- We need to have cicd for some mature pipeline (like speech), right now it's mostly manual
- Change to distillation in the future to mitigate the model drift effect
- We should have run this on 8xA6000 than the runpod
